### PR TITLE
Support ProductInfo-based IDEs

### PR DESCRIPTION
**Pull request by novotnyr**
- Support `product-info.json` for Platform plugins and module resolution
- Discover plugins and modules in the standard platform locations that are in the 2024.2+
- Support module descriptors in the `module-descriptor.jars`. Maintain classpaths and intra-module dependencies.
- Support reorganized plugins (Kotlin, Python) and their identifiers
- Establish a new `IdePluginManager` implementation that supports a new layout

### CODE REVIEW

File: intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/layout/PlatformPluginManager.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        META_INF / "${platformPrefix}Plugin.xml",
+        META_INF / PLUGIN_XML,
+        META_INF / PLATFORM_PLUGIN_XML
+      )
+    }
+
+  private val IdeVersion.platformPrefix: String
+    get() {
+      val platformProduct = IntelliJPlatformProduct.fromIdeVersion(this)
+      val product = platformProduct ?: IntelliJPlatformProduct.IDEA
```
Comment (YannCebron) [inline intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/layout/PlatformPluginManager.kt:62]: how safe is this? shouldn't we fail instead?

File: intellij-plugin-structure/structure-intellij/src/test/kotlin/com/jetbrains/plugin/structure/intellij/platform/ModuleBeanParsingTest.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.jetbrains.plugin.structure.intellij.beans.ModuleBean
+import com.jetbrains.plugin.structure.intellij.extractor.ModuleUnmarshaller
+import junit.framework.TestCase.assertNotNull
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertThrows
+import org.junit.Test
+import java.nio.file.Path
+import javax.xml.bind.UnmarshalException
+
+class ModuleBeanParsingTest {
```
Comment (YannCebron) [inline intellij-plugin-structure/structure-intellij/src/test/kotlin/com/jetbrains/plugin/structure/intellij/platform/ModuleBeanParsingTest.kt:12]: why do we rely on functionality that is still WIP and only available in 242+ AFAIK?

File: intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/ProductInfoBasedIdeManager.kt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    idePath: Path,
+    productInfo: ProductInfo,
+    ideVersion: IdeVersion
+  ): List<IdePlugin> {
+
+    val platformResourceResolver = getPlatformResourceResolver(productInfo, idePath)
+    val moduleManager = BundledModulesManager(BundledModulesResolver(idePath))
+
+    val productModuleV2Factory = ModuleFactory(::createProductModule, ProductInfoClasspathProvider(productInfo))
+    val moduleV2Factory = productModuleV2Factory
```
[the remark this answers is not in the document]
Comment (novotnyr) [reply intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/ProductInfoBasedIdeManager.kt]: Clarified the same mechanism in the code with the same factory object.

File: intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/layout/PlatformPluginManager.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import com.jetbrains.plugin.structure.jar.PluginDescriptorResult
+import com.jetbrains.plugin.structure.jar.PluginJar
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import java.io.File
+import java.nio.file.Path
+
+private val LOG: Logger = LoggerFactory.getLogger(PlatformPluginManager::class.java)
+
+internal class PlatformPluginManager(private val pluginLoader: LayoutComponentLoader) {
```
[the remark this answers is not in the document]
Comment (novotnyr) [reply intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/layout/PlatformPluginManager.kt:23]: Platform JARs reside in `IDE_HOME/lib` directory, as per [PlatformLayout](https://github.com/JetBrains/intellij-community/blob/e078d61f0345791bd9ea1d2c1b8bc9f8195fe0b2/platform/build-scripts/src/org/jetbrains/intellij/build/impl/PlatformLayout.kt).

Also, PV already contains [similar method](https://github.com/JetBrains/intellij-plugin-verifier/blob/c6e0c666d3bb03d1c95a4765eda95ff2f6672e23/intellij-plugin-structure/structure-ide/src/main/java/com/jetbrains/plugin/structure/ide/IdeManagerImpl.kt#L196) to indicate such plugins.

**Review by YannCebron — APPROVED** [review]

**Review by LChernigovskaya — APPROVED** [review]

File: intellij-plugin-structure/structure-intellij/src/test/kotlin/com/jetbrains/plugin/structure/intellij/platform/ModuleBeanParsingTest.kt (hunk 2 above)
Comment (novotnyr) [reply intellij-plugin-structure/structure-intellij/src/test/kotlin/com/jetbrains/plugin/structure/intellij/platform/ModuleBeanParsingTest.kt:12]: Using just the resource paths and module names. No dependency management is taken from this file.

**Pull request merged**